### PR TITLE
Exclude System.* reference assemblies in ILCompiler.Build.Tasks

### LINKED ISSUE #83695: Failure when building self-hosted NativeAOT compiler changes with 8.0 Preview 2 SDK
The changes in https://github.com/dotnet/runtime/pull/81205 have caused a failure when attempting to use .NET 8 Preview 2 SDK to build the Preview 3 source. This is a required scenario for .NET's Source-Build. This was identified in the changes to update the source-build configuration so that it references the Preview 2 SDK: https://github.com/dotnet/installer/pull/15851.

The error that occurs is for the ILCompiler project:

```
EXEC : error : Failed to load assembly 'System.CommandLine' [/vmr/src/runtime/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj]
      Internal.TypeSystem.TypeSystemException+FileNotFoundException: Failed to load assembly 'System.CommandLine'
         at Internal.TypeSystem.ThrowHelper.ThrowFileNotFoundException(ExceptionStringID, String) in /_/src/coreclr/tools/Common/TypeSystem/Common/ThrowHelper.cs:line 35
         at Internal.TypeSystem.ResolutionFailure.Throw() in /_/src/coreclr/tools/Common/TypeSystem/Common/ResolutionFailure.cs:line 105
         at Internal.TypeSystem.Ecma.EcmaModule.GetObject(EntityHandle, NotFoundBehavior ) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs:line 412
         at Internal.TypeSystem.Ecma.EcmaModule.GetType(EntityHandle) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaModule.cs:line 359
         at Internal.TypeSystem.Ecma.EcmaType.InitializeBaseType() in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:line 163
         at Internal.TypeSystem.Ecma.EcmaType.ComputeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:line 199
         at Internal.TypeSystem.TypeDesc.InitializeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Common/TypeDesc.cs:line 135
         at Internal.TypeSystem.Ecma.EcmaType.ComputeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs:line 247
         at Internal.TypeSystem.TypeDesc.InitializeTypeFlags(TypeFlags) in /_/src/coreclr/tools/Common/TypeSystem/Common/TypeDesc.cs:line 135
         at Internal.TypeSystem.MetadataFieldLayoutAlgorithm.ComputeInstanceLayout(DefType, InstanceLayoutKind) in /_/src/coreclr/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:line 41
         at Internal.TypeSystem.DefType.ComputeInstanceLayout(InstanceLayoutKind) in /_/src/coreclr/tools/Common/TypeSystem/Common/DefType.FieldLayout.cs:line 436
         at ILCompiler.CompilerTypeSystemContext.EnsureLoadableTypeUncached(TypeDesc) in /_/src/coreclr/tools/Common/Compiler/CompilerTypeSystemContext.Validation.cs:line 137
         at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
         at ILCompiler.DependencyAnalysis.EETypeNode..ctor(NodeFactory, TypeDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/EETypeNode.cs:line 85
         at ILCompiler.DependencyAnalysis.NodeFactory.CreateConstructedTypeNode(TypeDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:line 542
         at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) in /_/src/coreclr/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs:line 562
         at ILCompiler.DependencyAnalysis.NodeFactory.ConstructedTypeSymbol(TypeDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/NodeFactory.cs:line 614
         at ILCompiler.DependencyAnalysis.ReflectionInvokeMapNode.AddDependenciesDueToReflectability(DependencyList&, NodeFactory, MethodDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectionInvokeMapNode.cs:line 54
         at ILCompiler.MetadataManager.GetDependenciesDueToReflectability(DependencyList&, NodeFactory, MethodDesc) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/MetadataManager.cs:line 399
         at ILCompiler.DependencyAnalysis.ReflectedMethodNode.GetStaticDependencies(NodeFactory) in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ReflectedMethodNode.cs:line 39
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 182
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependencies(DependencyNodeCore`1) in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 222
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 257
         at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() in /_/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/DependencyAnalyzer.cs:line 308
         at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() in /_/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:line 140
         at ILCompiler.Program.<Run>g__RunScanner|3_0(<>c__DisplayClass3_0&) in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 438
         at ILCompiler.Program.Run() in /_/src/coreclr/tools/aot/ILCompiler/Program.cs:line 418
         at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass203_0.<.ctor>b__0(InvocationContext) in /_/src/coreclr/tools/aot/ILCompiler/ILCompilerRootCommand.cs:line 272
```

An analysis of the error and its symptoms is given here: https://github.com/dotnet/installer/pull/15851#issuecomment-1474447976. As stated in the analysis, the behavior is sporadic for some reason. One notable symptom is that [`BadImageFormatException` is thrown when attempting to load certain assemblies](https://github.com/dotnet/installer/pull/15851#issuecomment-1476671469) in the `ComputeManagedAssembliesToCompileToNative` build task.

It's been confirmed that reverting the changes in https://github.com/dotnet/runtime/pull/81205 resolves the issue. We should consider reverting this change or getting a fix in to address this in the Preview 3 timeframe.

cc @MichalStrehovsky

### PR DESCRIPTION
**Pull request by jkotas**
We expect the implementation of these assemblies to come as part of msbuild.

Fixes #83695

### CODE REVIEW

**Comment by jkotas** [comment]
It is not the most robust fix, but I cannot think about anything better that would not be too complicated.

**Review by agocke — APPROVED** [review]
I think this looks good, but I'm not sure why we're getting reference assemblies being copied (per your other comment). This is netstandard2.0, so I would expect that we would get either real assemblies for the non-inbox stuff, or facade assemblies. Why reference assemblies?

File: src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj [hunk 1]
```diff
@@ -16,6 +16,11 @@
     <PackageReference Include="Microsoft.Build.Framework" Version="$(MicrosoftBuildFrameworkVersion)" PrivateAssets="all" ExcludeAssets="runtime" />
     <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(MicrosoftBuildUtilitiesCoreVersion)" PrivateAssets="all" ExcludeAssets="runtime" />
     <PackageReference Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
```
Comment (agocke) [inline src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj:18]: Actually, SRM should be in the SDK as well. Can we add PrivateAssets here as well?

**Review by sbomer — APPROVED** [review]

**Comment by agocke** [comment]
OK, I think I'm even more confused after thinking about this more.

I previously said S.R.M is in the SDK. I think that's nominally correct -- the MSBuild.deps.json file lists S.R.M as a dependency.

But presumably if we're running the desktop MSBuild, that will not load the MSBuild from the SDK.

But if we're running desktop MSBuild I would presume that we do actually need things like System.Memory and System.Numerics.Vectors. Those are not in-box in 48, are they?

So this kind of leads me back to the original question: why are there ref assemblies of those DLLs in the output? Shouldn't those be actual DLLs that are loaded only on desktop FX?

Looking at Roslyn, they eventually moved to multi-targeting the task, for a variety of reasons, but one of them was this kind of confusion about what needed to be included in the package. I don't really have an opinion either way on whether we should go in that direction, but I find it notable that System.Memory and System.Numerics.Vectors do appear in the 472 version of their task (and are missing from the netcore version).

**Comment by jkotas** [comment]
> But presumably if we're running the desktop MSBuild, that will not load the MSBuild from the SDK.

Desktop MSBuild in supported VS versions come with these `System.*` assemblies. 
```
 Directory of c:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin

05/17/2023  10:57 AM            20,856 System.Buffers.dll
05/17/2023  10:57 AM           198,784 System.Collections.Immutable.dll
05/17/2023  10:57 AM           142,240 System.Memory.dll
05/17/2023  10:57 AM           115,856 System.Numerics.Vectors.dll
05/17/2023  10:57 AM           466,576 System.Reflection.Metadata.dll
05/17/2023  10:57 AM           245,888 System.Reflection.MetadataLoadContext.dll
05/17/2023  10:57 AM            61,568 System.Resources.Extensions.dll
05/17/2023  10:57 AM            18,024 System.Runtime.CompilerServices.Unsafe.dll
05/17/2023  10:57 AM            78,976 System.Text.Encodings.Web.dll
05/17/2023  10:57 AM           582,800 System.Text.Json.dll
05/17/2023  10:57 AM           181,376 System.Threading.Tasks.Dataflow.dll
05/17/2023  10:57 AM            25,984 System.Threading.Tasks.Extensions.dll
05/17/2023  10:57 AM            25,232 System.ValueTuple.dll
```

We should be able to delete `System.Reflection.Metadata`  and `System.Collections.Immutable` too. Other msbuild tasks in the repo assume that msbuild comes with good version of these assemblies (e.g. see conversation at https://github.com/dotnet/runtime/pull/85738#issuecomment-1537190426).

> why are there ref assemblies of those DLLs in the output? Shouldn't those be actual DLLs that are loaded only on desktop FX?

No idea. Publishing of libraries projects has rough edges. I count this as one of the rough edges.

> Looking at Roslyn, they eventually moved to multi-targeting the task, for a variety of reasons, but one of them was this kind of confusion about what needed to be included in the package. 

I do not think we need to do that. We have much smaller support matrix compared to Roslyn (last SDK only, VS versions that have support for the given SDK only), so the current scheme is good enough for us. Also, other tasks in the repo are on the same plan.

File: src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj (hunk 1 above)
Comment (jkotas) [reply src/coreclr/tools/aot/ILCompiler.Build.Tasks/ILCompiler.Build.Tasks.csproj:17]: Delete CopyLocalLockFileAssemblies property instead.

**Pull request merged**